File: commands/duel.js
const { SlashCommandBuilder, GuildMember } = require("discord.js");
const { newChallenge } = require("../game-manager.js");

module.exports = {
    data: new SlashCommandBuilder()
        .setName("duel")
        .setDescription("Begin a new game of wordle")
        .addMentionableOption(option => 
            option.setName("opponent")
                .setDescription("Your noble opponent")
                .setRequired(true),
        ),
    async execute(interaction){
        let opponent = interaction.options.getMentionable("opponent");
        if(!isValidOpponent(opponent))
            interaction.reply({content: "Please duel a valid user!", ephemeral: true});
        else {
            // A valid duel has been started
            if(newChallenge(interaction.user.id, opponent.user.id)){
                // Send the duel invite to the other player
                interaction.reply({content: "Challenge sent! Good luck.", ephemeral: true});
                interaction.channel.send(`${opponent.user}, ${interaction.user} has challenged you to a duel! Type \`/accept\` or \`/decline\` to respond!`);
            } else
                interaction.reply({content: "Each player can only have one active challenge!", ephemeral: true});
        }
    },
};

function isValidOpponent(opponent){
    return opponent instanceof GuildMember;
}

File: commands/duel.go
package commands

import (
	"fmt"

	"cordle/game"
	"cordle/config"

	"github.com/bwmarrin/discordgo"
)

func duel(s *discordgo.Session, i *discordgo.InteractionCreate){
	// TODO: check whether a role is mentioned
	
	// Check that the target does not already have a challenge against them
	user := i.ApplicationCommandData().Options[0].UserValue(s)
	if game.FindChallenge(user) == nil{
		// Create a new challenge
		game.NewChallenge(i.Interaction.Member.User, user)
		// Respond to the interaction, notifying the other player of the duel
		s.InteractionRespond(i.Interaction, &discordgo.InteractionResponse{
			Type: discordgo.InteractionResponseChannelMessageWithSource,
			Data: &discordgo.InteractionResponseData{
				Content: fmt.Sprintf(
					"%s, %s has challenged you to a duel! You have %d seconds to either `/accept` or `/decline` this duel",
					user.Mention(),
					i.Interaction.Member.Mention(),
					config.Config.Game.ChallengeDuration,
				),
			},
		})
	} else {
		// Challenge already exists
		ephemeralResponse(s, i, "That player already has an active challenge against them!")
	}
}